docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,871 @@
+
+Feel free to send us PRs against the `v6/environment-api` branch to fix the issues you discover. Please share with us your feedback as you test the proposal.
+:::
+
+A single Vite dev server can be used to interact with different module execution environments concurrently. We'll use the word environment to refer to a configured Vite processing pipeline that can resolve ids, load, and process source code and is connected to a runtime where the code is executed. The transformed source code is called a module, and the relationships between the modules processed in each environment are kept in a module graph. The code for these modules is sent to the runtimes associated with each environment to be executed. When a module is evaluated, the runtime will request its imported modules triggering the processing of a section of the module graph. In a typical Vite app, an environments will be used for the ES modules served to the client and for the server program that does SSR. An app can do SSR on a Node server, but also on an edge runtime like Workerd. So we can have different types of environments on the same Vite server: browser environments, node environments, and workerd environments, to name a few.
+
+A Vite Module Runner allows running any code by processing it with Vite plugins first. It is different from `server.ssrLoadModule` because the runner implementation is decoupled from the server. This allows library and framework authors to implement their layer of communication between the Vite server and the runner. The browser communicates with its corresponding environment using the server Web Socket and through HTTP requests. The Node Module runner can directly do function calls to process modules as it is running in the same process. Other environments could run modules connecting to an edge runtime like workerd, or a Worker Thread as Vitest does.
+
+All these environments share Vite's HTTP server, middlewares, and Web Socket. The resolved config and plugins pipeline are also shared, but plugins can use `apply` so its hooks are only called for certain environments. The environment can also be accessed inside hooks for fine-grained control.
+
+![Vite Environments](../images/vite-environments.svg)
+
+## Using environments in the Vite server
+
+A Vite dev server exposes two environments by default: a Client environment and a SSR environment. The client environment is a browser environment by default, and the module runner is implemented by importing the virtual module `/@vite/client` to client apps. The SSR environment runs in the same Node runtime as the Vite server by default and allows application servers to be used to render requests during dev with full HMR support. We'll discuss later how frameworks and users can change the environment types for the default client and SSR environments, or register new environments (for example to have a separate module graph for RSC).
```

Comment:
We discussed this in person with @patak-dev and he made an argument that browser might need multiple environments (e.g. for web worker) and there are also other non-browser clients, like electron.

My 2c is that we should name environments by the purpose they serve rather than technology that powers them or generic category like "client" that can be easily misunderstood. So following this logic, the default browser environment could be called browserMain or browserUI to represent the browser main thread or ui envionment.

Reply:
Calling `browserMain` to tauri doesn't seems to me to convey the purpose. I think the purpose of the browser, tauri, electron, web extension, etc is to be the `client` of the app. There is another reason for the `client` and `ssr` naming in that it has been the way we have always called these two environments (there are options prefixed by `ssr` and by `client` in the config currently)